src/MetaTags/CanonicalUrl.php

```diff
@@ -31,10 +30,25 @@ private function get_home_value() {
 		return home_url( '/' );
 	}
 
-	private function get_singular_value() {
+	private function get_singular_value( $post_id = null ) {
+		$post_id = $post_id ?: $this->get_queried_object_id();
+		$post    = get_post( $post_id );
+		if ( ! $post ) {
+			return '';
+		}
+
+		$data = get_post_meta( $post_id, 'slim_seo', true );
+		if ( ! empty( $data['canonical'] ) && wp_http_validate_url( $data['canonical'] ) ) {
+			return $data['canonical'];
+		}
+
```

Comment:
Hàm này ko cần phải truyền `$post_id` vào đâu em. Em lấy trực tiếp từ `$this->get_queried_object_id()` là ok rồi mà. Ngoài ra, ko cần phải validate URL đâu, cẩn thận quá :)